File: swh/web/admin/origin_save.py
# Copyright (C) 2018  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU Affero General Public License version 3, or any later version
# See top-level LICENSE file for more information

import json

from django.conf import settings
from django.contrib.admin.views.decorators import staff_member_required
from django.core.exceptions import ObjectDoesNotExist
from django.core.paginator import Paginator
from django.http import HttpResponse
from django.shortcuts import render
from django.views.decorators.http import require_POST

from swh.web.admin.adminurls import admin_route

from swh.web.common.models import (
    SaveAuthorizedOrigin, SaveUnauthorizedOrigin, SaveOriginRequest
)

from swh.web.common.origin_save import (
    create_save_origin_request,
    SAVE_REQUEST_PENDING, SAVE_REQUEST_REJECTED
)


@admin_route(r'origin/save/', view_name='admin-origin-save')
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save(request):
    return render(request, 'admin/origin-save.html')


def _datatables_origin_urls_response(request, urls_query_set):
    search_value = request.GET['search[value]']
    if search_value:
        urls_query_set = urls_query_set.filter(url__icontains=search_value)

    column_order = request.GET['order[0][column]']
    field_order = request.GET['columns[%s][name]' % column_order]
    order_dir = request.GET['order[0][dir]']
    if order_dir == 'desc':
        field_order = '-' + field_order

    urls_query_set = urls_query_set.order_by(field_order)

    table_data = {}
    table_data['draw'] = int(request.GET['draw'])
    table_data['recordsTotal'] = urls_query_set.count()
    table_data['recordsFiltered'] = urls_query_set.count()
    length = int(request.GET['length'])
    page = int(request.GET['start']) / length + 1
    paginator = Paginator(urls_query_set, length)
    urls_query_set = paginator.page(page).object_list
    table_data['data'] = [{'url': u.url} for u in urls_query_set]
    table_data_json = json.dumps(table_data, separators=(',', ': '))
    return HttpResponse(table_data_json, content_type='application/json')


@admin_route(r'origin/save/authorized_urls/list/',
             view_name='admin-origin-save-authorized-urls-list')
@staff_member_required
def _admin_origin_save_authorized_urls_list(request):
    authorized_urls = SaveAuthorizedOrigin.objects.all()
    return _datatables_origin_urls_response(request, authorized_urls)


@admin_route(r'origin/save/authorized_urls/add/(?P<origin_url>.+)/',
             view_name='admin-origin-save-add-authorized-url')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_add_authorized_url(request, origin_url):
    try:
        SaveAuthorizedOrigin.objects.get(url=origin_url)
    except ObjectDoesNotExist:
        # add the new authorized url
        SaveAuthorizedOrigin.objects.create(url=origin_url)
        # check if pending save requests with that url prefix exist
        pending_save_requests = \
            SaveOriginRequest.objects.filter(origin_url__startswith=origin_url,
                                             status=SAVE_REQUEST_PENDING)
        # create origin save tasks for previously pending requests
        for psr in pending_save_requests:
            create_save_origin_request(psr.origin_type, psr.origin_url)
        status_code = 200
    else:
        status_code = 400
    return HttpResponse(status=status_code)


@admin_route(r'origin/save/authorized_urls/remove/(?P<origin_url>.+)/',
             view_name='admin-origin-save-remove-authorized-url')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_remove_authorized_url(request, origin_url):
    try:
        entry = SaveAuthorizedOrigin.objects.get(url=origin_url)
    except ObjectDoesNotExist:
        status_code = 404
    else:
        entry.delete()
        status_code = 200
    return HttpResponse(status=status_code)


@admin_route(r'origin/save/unauthorized_urls/list/',
             view_name='admin-origin-save-unauthorized-urls-list')
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_unauthorized_urls_list(request):
    unauthorized_urls = SaveUnauthorizedOrigin.objects.all()
    return _datatables_origin_urls_response(request, unauthorized_urls)


@admin_route(r'origin/save/unauthorized_urls/add/(?P<origin_url>.+)/',
             view_name='admin-origin-save-add-unauthorized-url')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_add_unauthorized_url(request, origin_url):
    try:
        SaveUnauthorizedOrigin.objects.get(url=origin_url)
    except ObjectDoesNotExist:
        SaveUnauthorizedOrigin.objects.create(url=origin_url)
        # check if pending save requests with that url prefix exist
        pending_save_requests = \
            SaveOriginRequest.objects.filter(origin_url__startswith=origin_url,
                                             status=SAVE_REQUEST_PENDING)
        # mark pending requests as rejected
        for psr in pending_save_requests:
            psr.status = SAVE_REQUEST_REJECTED
            psr.save()
        status_code = 200
    else:
        status_code = 400
    return HttpResponse(status=status_code)


@admin_route(r'origin/save/unauthorized_urls/remove/(?P<origin_url>.+)/',
             view_name='admin-origin-save-remove-unauthorized-url')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_remove_unauthorized_url(request, origin_url):
    try:
        entry = SaveUnauthorizedOrigin.objects.get(url=origin_url)
    except ObjectDoesNotExist:
        status_code = 404
    else:
        entry.delete()
        status_code = 200
    return HttpResponse(status=status_code)


@admin_route(r'origin/save/request/accept/(?P<origin_type>.+)/url/(?P<origin_url>.+)/', # noqa
             view_name='admin-origin-save-request-accept')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_request_accept(request, origin_type, origin_url):
    SaveAuthorizedOrigin.objects.create(url=origin_url)
    create_save_origin_request(origin_type, origin_url)
    return HttpResponse(status=200)


@admin_route(r'origin/save/request/reject/(?P<origin_type>.+)/url/(?P<origin_url>.+)/', # noqa
             view_name='admin-origin-save-request-reject')
@require_POST
@staff_member_required(login_url=settings.LOGIN_URL)
def _admin_origin_save_request_reject(request, origin_type, origin_url):
    SaveUnauthorizedOrigin.objects.create(url=origin_url)
    sor = SaveOriginRequest.objects.get(origin_type=origin_type,
                                        origin_url=origin_url,
                                        status=SAVE_REQUEST_PENDING)
    sor.status = SAVE_REQUEST_REJECTED
    sor.save()
    return HttpResponse(status=200)

File: assets/src/bundles/admin/origin-save.js
/**
 * Copyright (C) 2018-2021  The Software Heritage developers
 * See the AUTHORS file at the top-level directory of this distribution
 * License: GNU Affero General Public License version 3, or any later version
 * See top-level LICENSE file for more information
 */

import {handleFetchError, csrfPost, htmlAlert} from 'utils/functions';
import {swhSpinnerSrc} from 'utils/constants';

let authorizedOriginTable;
let unauthorizedOriginTable;
let pendingSaveRequestsTable;
let acceptedSaveRequestsTable;
let rejectedSaveRequestsTable;

function enableRowSelection(tableSel) {
  $(`${tableSel} tbody`).on('click', 'tr', function() {
    if ($(this).hasClass('selected')) {
      $(this).removeClass('selected');
      $(tableSel).closest('.tab-pane').find('.swh-action-need-selection').prop('disabled', true);
    } else {
      $(`${tableSel} tr.selected`).removeClass('selected');
      $(this).addClass('selected');
      $(tableSel).closest('.tab-pane').find('.swh-action-need-selection').prop('disabled', false);
    }
  });
}

export function initOriginSaveAdmin() {
  $(document).ready(() => {

    $.fn.dataTable.ext.errMode = 'throw';

    authorizedOriginTable = $('#swh-authorized-origin-urls').DataTable({
      serverSide: true,
      ajax: Urls.admin_origin_save_authorized_urls_list(),
      columns: [{data: 'url', name: 'url'}],
      scrollY: '50vh',
      scrollCollapse: true,
      info: false
    });
    enableRowSelection('#swh-authorized-origin-urls');
    swh.webapp.addJumpToPagePopoverToDataTable(authorizedOriginTable);

    unauthorizedOriginTable = $('#swh-unauthorized-origin-urls').DataTable({
      serverSide: true,
      ajax: Urls.admin_origin_save_unauthorized_urls_list(),
      columns: [{data: 'url', name: 'url'}],
      scrollY: '50vh',
      scrollCollapse: true,
      info: false
    });
    enableRowSelection('#swh-unauthorized-origin-urls');
    swh.webapp.addJumpToPagePopoverToDataTable(unauthorizedOriginTable);

    const columnsData = [
      {
        data: 'id',
        name: 'id',
        visible: false,
        searchable: false
      },
      {
        data: 'save_request_date',
        name: 'request_date',
        render: (data, type, row) => {
          if (type === 'display') {
            const date = new Date(data);
            return date.toLocaleString();
          }
          return data;
        }
      },
      {
        data: 'visit_type',
        name: 'visit_type'
      },
      {
        data: 'origin_url',
        name: 'origin_url',
        render: (data, type, row) => {
          if (type === 'display') {
            let html = '';
            const sanitizedURL = $.fn.dataTable.render.text().display(data);
            if (row.save_task_status === 'succeeded') {
              let browseOriginUrl = `${Urls.browse_origin()}?origin_url=${encodeURIComponent(sanitizedURL)}`;
              if (row.visit_date) {
                browseOriginUrl += `&amp;timestamp=${encodeURIComponent(row.visit_date)}`;
              }
              html += `<a href="${browseOriginUrl}">${sanitizedURL}</a>`;
            } else {
              html += sanitizedURL;
            }
            html += `&nbsp;<a href="${sanitizedURL}" target="_blank" rel="noopener noreferrer">` +
              '<i class="mdi mdi-open-in-new" aria-hidden="true"></i></a>';
            return html;
          }
          return data;
        }
      }
    ];

    pendingSaveRequestsTable = $('#swh-origin-save-pending-requests').DataTable({
      serverSide: true,
      processing: true,
      language: {
        processing: `<img src="${swhSpinnerSrc}"></img>`
      },
      ajax: Urls.origin_save_requests_list('pending'),
      searchDelay: 1000,
      columns: columnsData,
      scrollY: '50vh',
      scrollCollapse: true,
      order: [[0, 'desc']],
      responsive: {
        details: {
          type: 'none'
        }
      }
    });
    enableRowSelection('#swh-origin-save-pending-requests');
    swh.webapp.addJumpToPagePopoverToDataTable(pendingSaveRequestsTable);

    columnsData.push({
      name: 'info',
      render: (data, type, row) => {
        if (row.save_task_status === 'succeeded' || row.save_task_status === 'failed' ||
            row.note != null) {
          return `<i class="mdi mdi-information-outline swh-save-request-info" aria-hidden="true"
                     style="cursor: pointer"
                     onclick="swh.save.displaySaveRequestInfo(event, ${row.id})"></i>`;
        } else {
          return '';
        }
      }
    });

    rejectedSaveRequestsTable = $('#swh-origin-save-rejected-requests').DataTable({
      serverSide: true,
      processing: true,
      language: {
        processing: `<img src="${swhSpinnerSrc}"></img>`
      },
      ajax: Urls.origin_save_requests_list('rejected'),
      searchDelay: 1000,
      columns: columnsData,
      scrollY: '50vh',
      scrollCollapse: true,
      order: [[0, 'desc']],
      responsive: {
        details: {
          type: 'none'
        }
      }
    });
    enableRowSelection('#swh-origin-save-rejected-requests');
    swh.webapp.addJumpToPagePopoverToDataTable(rejectedSaveRequestsTable);

    columnsData.splice(columnsData.length - 1, 0, {
      data: 'save_task_status',
      name: 'save_task_status'
    });

    acceptedSaveRequestsTable = $('#swh-origin-save-accepted-requests').DataTable({
      serverSide: true,
      processing: true,
      language: {
        processing: `<img src="${swhSpinnerSrc}"></img>`
      },
      ajax: Urls.origin_save_requests_list('accepted'),
      searchDelay: 1000,
      columns: columnsData,
      scrollY: '50vh',
      scrollCollapse: true,
      order: [[0, 'desc']],
      responsive: {
        details: {
          type: 'none'
        }
      }
    });
    enableRowSelection('#swh-origin-save-accepted-requests');
    swh.webapp.addJumpToPagePopoverToDataTable(acceptedSaveRequestsTable);

    $('#swh-origin-save-requests-nav-item').on('shown.bs.tab', () => {
      pendingSaveRequestsTable.draw();
    });

    $('#swh-origin-save-url-filters-nav-item').on('shown.bs.tab', () => {
      authorizedOriginTable.draw();
    });

    $('#swh-authorized-origins-tab').on('shown.bs.tab', () => {
      authorizedOriginTable.draw();
    });

    $('#swh-unauthorized-origins-tab').on('shown.bs.tab', () => {
      unauthorizedOriginTable.draw();
    });

    $('#swh-save-requests-pending-tab').on('shown.bs.tab', () => {
      pendingSaveRequestsTable.draw();
    });

    $('#swh-save-requests-accepted-tab').on('shown.bs.tab', () => {
      acceptedSaveRequestsTable.draw();
    });

    $('#swh-save-requests-rejected-tab').on('shown.bs.tab', () => {
      rejectedSaveRequestsTable.draw();
    });

    $('#swh-save-requests-pending-tab').click(() => {
      pendingSaveRequestsTable.ajax.reload(null, false);
    });

    $('#swh-save-requests-accepted-tab').click(() => {
      acceptedSaveRequestsTable.ajax.reload(null, false);
    });

    $('#swh-save-requests-rejected-tab').click(() => {
      rejectedSaveRequestsTable.ajax.reload(null, false);
    });

    $('body').on('click', e => {
      if ($(e.target).parents('.popover').length > 0) {
        e.stopPropagation();
      } else if ($(e.target).parents('.swh-save-request-info').length === 0) {
        $('.swh-save-request-info').popover('dispose');
      }
    });

  });
}

export async function addAuthorizedOriginUrl() {
  const originUrl = $('#swh-authorized-url-prefix').val();
  const addOriginUrl = Urls.admin_origin_save_add_authorized_url(originUrl);
  try {
    const response = await csrfPost(addOriginUrl);
    handleFetchError(response);
    authorizedOriginTable.row.add({'url': originUrl}).draw();
    $('.swh-add-authorized-origin-status').html(
      htmlAlert('success', 'The origin url prefix has been successfully added in the authorized list.', true)
    );
  } catch (_) {
    $('.swh-add-authorized-origin-status').html(
      htmlAlert('warning', 'The provided origin url prefix is already registered in the authorized list.', true)
    );
  }
}

export async function removeAuthorizedOriginUrl() {
  const originUrl = $('#swh-authorized-origin-urls tr.selected').text();
  if (originUrl) {
    const removeOriginUrl = Urls.admin_origin_save_remove_authorized_url(originUrl);
    try {
      const response = await csrfPost(removeOriginUrl);
      handleFetchError(response);
      authorizedOriginTable.row('.selected').remove().draw();
    } catch (_) {}
  }
}

export async function addUnauthorizedOriginUrl() {
  const originUrl = $('#swh-unauthorized-url-prefix').val();
  const addOriginUrl = Urls.admin_origin_save_add_unauthorized_url(originUrl);
  try {
    const response = await csrfPost(addOriginUrl);
    handleFetchError(response);
    unauthorizedOriginTable.row.add({'url': originUrl}).draw();
    $('.swh-add-unauthorized-origin-status').html(
      htmlAlert('success', 'The origin url prefix has been successfully added in the unauthorized list.', true)
    );
  } catch (_) {
    $('.swh-add-unauthorized-origin-status').html(
      htmlAlert('warning', 'The provided origin url prefix is already registered in the unauthorized list.', true)
    );
  }
}

export async function removeUnauthorizedOriginUrl() {
  const originUrl = $('#swh-unauthorized-origin-urls tr.selected').text();
  if (originUrl) {
    const removeOriginUrl = Urls.admin_origin_save_remove_unauthorized_url(originUrl);
    try {
      const response = await csrfPost(removeOriginUrl);
      handleFetchError(response);
      unauthorizedOriginTable.row('.selected').remove().draw();
    } catch (_) {};
  }
}

export function acceptOriginSaveRequest() {
  const selectedRow = pendingSaveRequestsTable.row('.selected');
  if (selectedRow.length) {
    const acceptOriginSaveRequestCallback = async() => {
      const rowData = selectedRow.data();
      const acceptSaveRequestUrl = Urls.admin_origin_save_request_accept(rowData['visit_type'], rowData['origin_url']);
      await csrfPost(acceptSaveRequestUrl);
      pendingSaveRequestsTable.ajax.reload(null, false);
    };

    swh.webapp.showModalConfirm(
      'Accept origin save request ?',
      'Are you sure to accept this origin save request ?',
      acceptOriginSaveRequestCallback);
  }
}

const rejectModalHtml = `
<form id="swh-rejection-form">
  <div class="form-group row">
    <label for="swh-rejection-reason" class="col-4 col-form-label">
      Rejection reason:
    </label>
    <div class="col-8">
      <select class="custom-select" id="swh-rejection-reason">
        <option value="custom" selected>Custom</option>
        <option value="invalid-origin">Invalid origin</option>
        <option value="invalid-origin-type">Invalid origin type</option>
        <option value="origin-not-found">Origin not found</option>
      </select>
    </div>
  </div>
  <div class="form-group row">
    <textarea class="form-control" id="swh-rejection-text"></textarea>
  </div>
  <button type="submit" class="btn btn-default float-right" id="swh-rejection-submit">
    Reject
  </button>
</form>
`;

export function rejectOriginSaveRequest() {
  const selectedRow = pendingSaveRequestsTable.row('.selected');
  const rowData = selectedRow.data();
  if (selectedRow.length) {
    const rejectOriginSaveRequestCallback = async() => {
      $('#swh-web-modal-html').modal('hide');
      const rejectSaveRequestUrl = Urls.admin_origin_save_request_reject(
        rowData['visit_type'], rowData['origin_url']);
      await csrfPost(rejectSaveRequestUrl, {},
                     JSON.stringify({note: $('#swh-rejection-text').val()}));
      pendingSaveRequestsTable.ajax.reload(null, false);
    };

    let currentRejectionReason = 'custom';
    const rejectionTexts = {};
    swh.webapp.showModalHtml('Reject origin save request ?', rejectModalHtml);
    $('#swh-rejection-reason').on('change', (event) => {
      // backup current textarea value
      rejectionTexts[currentRejectionReason] = $('#swh-rejection-text').val();
      currentRejectionReason = event.target.value;
      let newRejectionText = '';
      if (rejectionTexts.hasOwnProperty(currentRejectionReason)) {
        // restore previous textarea value
        newRejectionText = rejectionTexts[currentRejectionReason];
      } else {
        // fill textarea with default text according to rejection type
        if (currentRejectionReason === 'invalid-origin') {
          newRejectionText = `The origin with URL ${rowData['origin_url']} is not ` +
            `a link to a  ${rowData['visit_type']} repository.`;
        } else if (currentRejectionReason === 'invalid-origin-type') {
          newRejectionText = `The origin with URL ${rowData['origin_url']} is not ` +
            `of type ${rowData['visit_type']}.`;
        } else if (currentRejectionReason === 'origin-not-found') {
          newRejectionText = `The origin with URL ${rowData['origin_url']} cannot be found.`;
        }
      }
      $('#swh-rejection-text').val(newRejectionText);
    });
    $('#swh-rejection-form').on('submit', (event) => {
      event.preventDefault();
      event.stopPropagation();
      // ensure confirmation modal will be displayed above the html modal
      $('#swh-web-modal-html').css('z-index', 4000);
      swh.webapp.showModalConfirm(
        'Reject origin save request ?',
        'Are you sure to reject this origin save request ?',
        rejectOriginSaveRequestCallback);
    });
  }
}

function removeOriginSaveRequest(requestTable) {
  const selectedRow = requestTable.row('.selected');
  if (selectedRow.length) {
    const requestId = selectedRow.data()['id'];
    const removeOriginSaveRequestCallback = async() => {
      const removeSaveRequestUrl = Urls.admin_origin_save_request_remove(requestId);
      await csrfPost(removeSaveRequestUrl);
      requestTable.ajax.reload(null, false);
    };

    swh.webapp.showModalConfirm(
      'Remove origin save request ?',
      'Are you sure to remove this origin save request ?',
      removeOriginSaveRequestCallback);
  }
}

export function removePendingOriginSaveRequest() {
  removeOriginSaveRequest(pendingSaveRequestsTable);
}

export function removeAcceptedOriginSaveRequest() {
  removeOriginSaveRequest(acceptedSaveRequestsTable);
}

export function removeRejectedOriginSaveRequest() {
  removeOriginSaveRequest(rejectedSaveRequestsTable);
}
